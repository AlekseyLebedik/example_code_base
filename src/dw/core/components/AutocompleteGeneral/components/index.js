import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { components } from 'react-select';

import Typography from '@material-ui/core/Typography';
import TextField from '@material-ui/core/TextField';
import Popper from '@material-ui/core/Popper';
import Paper from '@material-ui/core/Paper';
import Chip from '@material-ui/core/Chip';
import MenuItem from '@material-ui/core/MenuItem';
import Icon from '@material-ui/core/Icon';

import styles from '../index.module.css';

function NoOptionsMessage(props) {
  const {
    selectProps: { classes, isAsync, inputValue },
    children,
    innerProps,
  } = props;
  return (
    <Typography
      color="textSecondary"
      className={classes.noOptionsMessage}
      {...innerProps}
    >
      {isAsync && !inputValue ? 'Start typing ...' : children}
    </Typography>
  );
}

NoOptionsMessage.propTypes = {
  selectProps: PropTypes.object.isRequired,
  innerProps: PropTypes.object,
  children: PropTypes.node.isRequired,
};

NoOptionsMessage.defaultProps = {
  innerProps: {},
};

function inputComponent({ inputRef, ...props }) {
  return <div ref={inputRef} {...props} />;
}

inputComponent.propTypes = {
  inputRef: PropTypes.func,
};

inputComponent.defaultProps = {
  inputRef: undefined,
};

function Control(props) {
  const {
    selectProps: { inputRef, classes, variant, size, textFieldProps, isMulti },
    children,
  } = props;
  return (
    <TextField
      fullWidth
      InputProps={{
        inputComponent,
        inputProps: {
          className: classNames(styles.textField, {
            [styles.bgWhite]: variant === 'box',
            [styles.slimSize]: size === 'slim',
            [styles.normalSize]: size === 'normal',
            [styles.bigSize]: size === 'big',
            [styles.textFieldMultiple]: isMulti,
          }),
          inputRef: ref => {
            props.innerRef(ref);
            inputRef(ref);
          },
          children,
          ...props.innerProps,
        },
        classes: {
          underline: classNames({
            [classes.underline]: variant === 'contrast',
          }),
        },
      }}
      {...textFieldProps}
    />
  );
}

Control.propTypes = {
  innerRef: PropTypes.func,
  selectProps: PropTypes.object.isRequired,
  innerProps: PropTypes.object,
  children: PropTypes.node.isRequired,
};

Control.defaultProps = {
  innerRef: undefined,
  innerProps: {},
};

function Option(props) {
  return (
    <MenuItem
      buttonRef={props.innerRef}
      selected={props.isFocused}
      disabled={props.isDisabled}
      component="div"
      style={{
        fontWeight: props.isSelected ? 600 : 400,
      }}
      data-cy="autocompleteOption"
      {...props.innerProps}
    >
      {props.children}
    </MenuItem>
  );
}

Option.propTypes = {
  innerRef: PropTypes.func,
  isFocused: PropTypes.bool,
  isDisabled: PropTypes.bool,
  isSelected: PropTypes.bool,
  innerProps: PropTypes.object,
  children: PropTypes.node.isRequired,
};

Option.defaultProps = {
  innerRef: undefined,
  isFocused: false,
  isDisabled: false,
  isSelected: false,
  innerProps: {},
};

function Placeholder(props) {
  const {
    selectProps: { classes, variant, fontSize },
  } = props;
  return (
    <Typography
      color="textSecondary"
      className={classNames(styles.placeholder, classes.placeholder, {
        [styles.white]: variant === 'contrast',
        [styles.withMargin]: variant === 'box',
        [styles.slimFontSize]: fontSize === 'slim',
        [styles.normalFontSize]: fontSize === 'normal',
        [styles.bigFontSize]: fontSize === 'big',
      })}
      {...props.innerProps}
    >
      {props.children}
    </Typography>
  );
}

Placeholder.propTypes = {
  selectProps: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired,
  innerProps: PropTypes.object,
};

Placeholder.defaultProps = {
  innerProps: {},
};

function SingleValue(props) {
  const {
    selectProps: { variant, fontSize, autoEllipsis },
  } = props;
  return (
    <Typography
      component="span"
      variant="body1"
      className={classNames(styles.singleValue, {
        [styles.white]: variant === 'contrast',
        [styles.withMargin]: variant === 'box',
        [styles.slimFontSize]: fontSize === 'slim',
        [styles.normalFontSize]: fontSize === 'normal',
        [styles.bigFontSize]: fontSize === 'big',
        [styles.autoEllipsis]: autoEllipsis,
      })}
      {...props.innerProps}
    >
      {props.children}
    </Typography>
  );
}

SingleValue.propTypes = {
  selectProps: PropTypes.object.isRequired,
  innerProps: PropTypes.object,
  children: PropTypes.node.isRequired,
};

SingleValue.defaultProps = {
  innerProps: {},
};

function ValueContainer(props) {
  const { children, isMulti, selectProps } = props;
  return (
    <div
      className={classNames(
        styles.valueContainer,
        selectProps.valueContainerClass,
        {
          [styles.valueContainerMultiple]: isMulti,
        }
      )}
    >
      {children}
    </div>
  );
}

ValueContainer.propTypes = {
  children: PropTypes.node.isRequired,
  isMulti: PropTypes.bool.isRequired,
  selectProps: PropTypes.object.isRequired,
};

function MultiValue(props) {
  return (
    <Chip
      tabIndex={-1}
      label={props.children}
      className={classNames(props.selectProps.classes.chip, {
        [props.selectProps.classes.chipFocused]: props.isFocused,
      })}
      onDelete={props.removeProps.onClick}
      deleteIcon={<Icon {...props.removeProps}>cancel</Icon>}
    />
  );
}

MultiValue.propTypes = {
  isFocused: PropTypes.bool,
  selectProps: PropTypes.object.isRequired,
  removeProps: PropTypes.object.isRequired,
  children: PropTypes.node.isRequired,
};

MultiValue.defaultProps = {
  isFocused: false,
};

function Menu(props) {
  return (
    <Popper
      key={props.selectProps.inputValue}
      anchorEl={props.selectProps.anchorEl}
      open={props.selectProps.menuIsOpen}
      className={props.selectProps.classes.popper}
    >
      <Paper
        square
        style={{
          width: props.selectProps.anchorEl
            ? props.selectProps.anchorEl.clientWidth
            : null,
        }}
        {...props.innerProps}
      >
        {props.children}
      </Paper>
    </Popper>
  );
}

Menu.propTypes = {
  selectProps: PropTypes.object.isRequired,
  innerProps: PropTypes.object,
  children: PropTypes.node.isRequired,
};

Menu.defaultProps = {
  innerProps: {},
};

function DropdownIndicator(props) {
  return (
    <components.DropdownIndicator {...props}>
      <Icon>arrow_drop_down</Icon>
    </components.DropdownIndicator>
  );
}

export const selectComponents = {
  Control,
  DropdownIndicator,
  Menu,
  MultiValue,
  NoOptionsMessage,
  Option,
  Placeholder,
  SingleValue,
  ValueContainer,
  IndicatorSeparator: () => null,
};
