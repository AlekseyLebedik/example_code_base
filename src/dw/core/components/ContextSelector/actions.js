import * as GlobalSnackBarActions from '@demonware/devzone-core/components/GlobalSnackBar/actions';
import {
  AVAILABLE_CONTEXTS_PREFIX,
  CONTEXT_REGISTRY_PREFIX,
  SERVICE_CHANGE_CONTEXT,
  CONTEXT_SELECTOR_CHANGE_USER,
  CONTEXT_SELECTOR_CHANGE_PROPS,
  CONTEXT_SELECTOR_RESET_PROPS,
} from './constants';

const prefixes = [AVAILABLE_CONTEXTS_PREFIX, CONTEXT_REGISTRY_PREFIX];

const createContextFetch = prefix => params => ({
  type: `${prefix}_FETCH`,
  params,
});

export const fetchAvailableContexts = createContextFetch(
  AVAILABLE_CONTEXTS_PREFIX
);

export const fetchContextsRegistry = createContextFetch(
  CONTEXT_REGISTRY_PREFIX
);

export const fetchSuccess = prefixes.reduce(
  (acc, prefix) => ({
    ...acc,
    [prefix]: (payload, { serviceName, userId }) => ({
      type: `${prefix}_FETCH_SUCCESS`,
      data: payload.data.data,
      serviceName,
      userId,
    }),
  }),
  {}
);

export const changeContext = (context, serviceName) => ({
  type: SERVICE_CHANGE_CONTEXT,
  context,
  serviceName,
});

export const changeUser = userId => ({
  type: CONTEXT_SELECTOR_CHANGE_USER,
  userId,
});

export const changeRemoteProps = props => ({
  type: CONTEXT_SELECTOR_CHANGE_PROPS,
  props,
});

export const resetRemoteProps = () => ({
  type: CONTEXT_SELECTOR_RESET_PROPS,
});

export const showNoAvailablePlatformContextMessage = serviceName =>
  GlobalSnackBarActions.show(
    `No Platform information available for this user.
    Ensure this user has a linked first party account and the ${serviceName} service is correctly configured`,
    'error'
  );

export const showNoAvailableContextMessage = serviceName =>
  GlobalSnackBarActions.show(
    `There are no available contexts for ${serviceName} service.
    Ensure the ${serviceName} service is correctly configured.`,
    'error'
  );
