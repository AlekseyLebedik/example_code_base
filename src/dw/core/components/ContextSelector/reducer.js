import get from 'lodash/get';
import {
  AVAILABLE_CONTEXTS_PREFIX,
  CONTEXT_REGISTRY_PREFIX,
  SERVICE_CHANGE_CONTEXT,
  CONTEXT_SELECTOR_CHANGE_USER,
  CONTEXT_SELECTOR_CHANGE_PROPS,
  CONTEXT_SELECTOR_RESET_PROPS,
} from './constants';

const initialState = {
  Available: {},
  Registry: {},
  User: { userId: null },
  Props: {},
  UserAvailableLoaded: false,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case `${AVAILABLE_CONTEXTS_PREFIX}_FETCH`:
      return {
        ...state,
        UserAvailableLoaded: false,
        userId: action.userId,
      };
    case `${AVAILABLE_CONTEXTS_PREFIX}_FETCH_SUCCESS`:
      return {
        ...state,
        Available: {
          ...state.Available,
          [action.serviceName]: {
            ...get(state, `Available.${action.serviceName}`, {}),
            data: action.data,
            currentContext: {
              // For each context type, find previous context in new available list and if matches, use previous context
              ...Object.keys(
                get(state, `Available.${action.serviceName}.currentContext`, {})
              ).map(type => {
                const currentContextName = get(
                  state,
                  `Available.${action.serviceName}.currentContext.${type}.name`
                );
                if (currentContextName) {
                  return action.data.find(
                    ctx => ctx.name === currentContextName
                  );
                }
                return undefined;
              }),
            },
          },
        },
        UserAvailableLoaded: !!action.userId,
        userId: action.userId,
      };
    case `${CONTEXT_REGISTRY_PREFIX}_FETCH_SUCCESS`:
      return {
        ...state,
        Registry: {
          ...state.Registry,
          [action.serviceName]: {
            ...state.Registry[action.serviceName],
            data: action.data,
          },
        },
      };
    case SERVICE_CHANGE_CONTEXT:
      return {
        ...state,
        Available: {
          ...state.Available,
          [action.serviceName]: {
            ...get(state, `Available.${action.serviceName}`, {}),
            currentContext: {
              [action.context.type]: action.context,
            },
          },
        },
      };
    case CONTEXT_SELECTOR_CHANGE_USER:
      return {
        ...state,
        User: { userId: action.userId },
      };
    case CONTEXT_SELECTOR_CHANGE_PROPS:
      return {
        ...state,
        Props: { ...action.props },
      };
    case CONTEXT_SELECTOR_RESET_PROPS:
      return {
        ...state,
        Props: {},
      };
    default:
      return state;
  }
};

export default reducer;
