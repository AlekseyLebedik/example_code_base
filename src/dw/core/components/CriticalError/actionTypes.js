export const CRITICAL_ERROR_SHOW = 'CRITICAL_ERROR_SHOW';
export const CRITICAL_ERROR_HIDE = 'CRITICAL_ERROR_HIDE';
