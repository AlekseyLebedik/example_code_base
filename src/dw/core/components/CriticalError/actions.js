import { CRITICAL_ERROR_SHOW, CRITICAL_ERROR_HIDE } from './actionTypes';

export const show = (error, retry) => ({
  type: CRITICAL_ERROR_SHOW,
  error,
  retry,
});

export const hide = () => ({
  type: CRITICAL_ERROR_HIDE,
});
