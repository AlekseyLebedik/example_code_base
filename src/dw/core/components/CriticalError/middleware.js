import { actions as GlobalProgressActions } from '@demonware/devzone-core/components/GlobalProgress';

import { CRITICAL_ERROR_SHOW } from './actionTypes';

export const middleware = store => next => action => {
  switch (action.type) {
    case CRITICAL_ERROR_SHOW:
      store.dispatch(GlobalProgressActions.done());
      break;
    default:
      break;
  }
  next(action);
};
