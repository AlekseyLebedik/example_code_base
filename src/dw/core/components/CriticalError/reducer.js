import { CRITICAL_ERROR_SHOW, CRITICAL_ERROR_HIDE } from './actionTypes';

const INITIAL_STATE = {
  error: undefined,
  retry: undefined,
};

export default function reducer(state = INITIAL_STATE, action) {
  switch (action.type) {
    case CRITICAL_ERROR_SHOW:
      return {
        ...state,
        error: action.error,
        retry: action.retry,
      };
    case CRITICAL_ERROR_HIDE:
      return INITIAL_STATE;
    default:
      return state;
  }
}
