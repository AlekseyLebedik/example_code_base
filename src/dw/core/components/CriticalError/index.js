import { Error500 } from './presentational';
import * as CriticalErrorActions from './actions';
import * as middleware from './middleware';

export { default as reducer } from './reducer';

export { default } from './container';

export { CriticalErrorActions, Error500, middleware };
