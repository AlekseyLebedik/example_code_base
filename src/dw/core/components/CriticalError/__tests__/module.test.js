import { actions as GlobalProgressActions } from '@demonware/devzone-core/components/GlobalProgress';

import reducer from '../reducer';
import { middleware } from '../middleware';
import * as actions from '../actions';
import * as AT from '../actionTypes';

describe('CriticalError', () => {
  let showErrorAction = null;
  let hideErrorAction = null;

  beforeEach(() => {
    showErrorAction = actions.show({ message: 'Error message' }, jest.fn());
    hideErrorAction = actions.hide();
  });

  describe('Action Creators', () => {
    it('CRITICAL_ERROR_SHOW', () => {
      expect(showErrorAction).toHaveProperty('type', AT.CRITICAL_ERROR_SHOW);
    });

    it('CRITICAL_ERROR_HIDE', () => {
      expect(hideErrorAction).toHaveProperty('type', AT.CRITICAL_ERROR_HIDE);
    });
  });

  describe('Reducer', () => {
    it('returns initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot();
    });

    it('handles CRITICAL_ERROR_SHOW', () => {
      expect(reducer(undefined, showErrorAction)).toMatchSnapshot();
    });

    it('handles CRITICAL_ERROR_HIDE', () => {
      expect(reducer(undefined, hideErrorAction)).toMatchSnapshot();

      const state = reducer(undefined, showErrorAction);
      expect(reducer(state, hideErrorAction)).toMatchSnapshot();
    });
  });

  describe('Middleware', () => {
    it('stops the global progress bar on every CRITICAL_ERROR_SHOW', () => {
      const storeMock = {
        getState: jest.fn(),
        dispatch: jest.fn(),
      };
      const next = jest.fn();

      middleware(storeMock)(next)(showErrorAction);

      expect(storeMock.dispatch).toHaveBeenCalledWith(
        GlobalProgressActions.done()
      );
      expect(next).toHaveBeenCalledWith(showErrorAction);
    });
  });
});
