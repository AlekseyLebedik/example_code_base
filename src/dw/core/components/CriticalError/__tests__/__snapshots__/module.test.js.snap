// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CriticalError Reducer handles CRITICAL_ERROR_HIDE 1`] = `
Object {
  "error": undefined,
  "retry": undefined,
}
`;

exports[`CriticalError Reducer handles CRITICAL_ERROR_HIDE 2`] = `
Object {
  "error": undefined,
  "retry": undefined,
}
`;

exports[`CriticalError Reducer handles CRITICAL_ERROR_SHOW 1`] = `
Object {
  "error": Object {
    "message": "Error message",
  },
  "retry": [MockFunction],
}
`;

exports[`CriticalError Reducer returns initial state 1`] = `
Object {
  "error": undefined,
  "retry": undefined,
}
`;
